Initial gaps support

stapelberg opened the pull request:
I still have a few refactorings I want to give a try before we should merge this, but I think the current state is good enough for a first pull request review :) 

related to https://github.com/i3/i3/issues/3724

stapelberg committed:
remove dead code: window_rect is overwritten a few lines below

stapelberg committed:
refactor render_con() global parameter into should_inset_con()

This bundles the logic all in one place and thereby makes it a little easier to
understand.

stapelberg committed:
Merge gaps support as-is

This code was copied over unmodified from https://github.com/Airblader/i3-gaps.

I have split out the differences between i3-gaps and i3 into three areas:
1. Gaps
2. i3bar height
3. rgba colors

stapelberg committed:
Fix config.spec comment

stapelberg committed:
gaps: use logical_px() to work correctly on hi-dpi monitors

stapelberg committed:
userguide: document gaps config directive and gaps command

stapelberg committed:
add basic gaps test

stapelberg committed:
gaps: allow optional px suffix

stapelberg committed:
move gaps-specific logic out of con.c and render.c into gaps.c

Airblader commented on docs/userguide:
```diff
@@ -1426,6 +1426,95 @@ tiling_drag modifier titlebar
+
+# Enabling “Smart Gaps” means no gaps will be shown when there is
+# precisely one window or split container on the workspace.
+smart_gaps on|off|inverse_outer
```
I think we need to explain the difference between on and inverse_outer here.

stapelberg replied:
Done

Airblader commented on parser-specs/config.spec:
```diff
@@ -65,6 +68,32 @@ state IGNORE_LINE:
+  enabled = 'no_gaps'
+      -> call cfg_smart_borders($enabled)
+
+# smart_gaps on|off
```
This comment needs to be updated, too.

stapelberg replied:
Done

stapelberg committed:
t/319-gaps.t: also test the gaps command

stapelberg committed:
refactor cmd_gaps to get rid of all #define

I’m still not 100% happy with how the function turned out (it still does too
many things at once), but this seems like an improvement — at least reading and
navigating the code with LSP now works better.

stapelberg committed:
config.spec: add missing smart_gaps inverse_outer comment

stapelberg commented:
Pushed the other commits I had in mind: moving code to gaps.c where applicable, allowing optional “px” suffix, getting rid of `#define` in cmd_gaps and test cases for the `gaps` command.

Please take another look. If this looks reasonable, we could merge this state and do further cleanups/improvements in separate PRs :)

Airblader commented on src/commands.c:
```diff
@@ -2375,114 +2375,171 @@ void cmd_debuglog(I3_CMD, const char *argument) {
+    /* See https://github.com/Airblader/i3/issues/262 */
+    int min_value = 0;
+    const bool is_outer = get(&(config.gaps)) != gaps_inner(&(config.gaps));
+    // if (strcmp(type, "inner") != 0) {
```
Leftover commented out code?

stapelberg replied:
Done

Airblader reviewed (approved):
LGTM % one nit. I have to admit that I find the define-based implementation easier to follow, but I might be biased there, am not really as experienced in C as my involvement in i3 would suggest, and it's not a big difference either way so happy to roll with this.

Thanks so much for taking care of this. We'll make a lot of people happy.
